### PR TITLE
Patches for v105

### PR DESCRIPTION
**Pull request by csagan5**
## Description

Patches for v105

## All submissions

* [x] there are no other open [Pull Requests](../../../pulls) for the same update/change
* [x] Bromite can be built with these changes
* [x] I have tested that the new change works as intended (AVD or physical device will do)

### Format

* [x] patch subject and filename match (e.g. `Subject: Alternative cache (NIK-based)` -> `Alternative-cache-NIK-based.patch`)
* [x] patch description contains explanation of changes
* [x] no unnecessary whitespace or unrelated changes

### CODE REVIEW

**Comment by uazo** [comment]
these are some fixups if you want to introduce them in the next build:

- Multiple fingerprinting mitigations (fixup)
fix the lack of protection in `canvas.convertToBlob()` 

```
From: Your Name <you@example.com>
Date: Fri, 2 Sep 2022 13:29:33 +0000
Subject: Multiple fingerprinting mitigations (fixup)

---
 .../renderer/core/html/canvas/canvas_async_blob_creator.cc   | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/third_party/blink/renderer/core/html/canvas/canvas_async_blob_creator.cc b/third_party/blink/renderer/core/html/canvas/canvas_async_blob_creator.cc
--- a/third_party/blink/renderer/core/html/canvas/canvas_async_blob_creator.cc
+++ b/third_party/blink/renderer/core/html/canvas/canvas_async_blob_creator.cc
@@ -20,6 +20,7 @@
 #include "third_party/blink/renderer/core/fileapi/blob.h"
 #include "third_party/blink/renderer/core/html/canvas/canvas_rendering_context.h"
 #include "third_party/blink/renderer/platform/graphics/image_data_buffer.h"
+#include "third_party/blink/renderer/platform/graphics/static_bitmap_image.h"
 #include "third_party/blink/renderer/platform/graphics/skia/skia_utils.h"
 #include "third_party/blink/renderer/platform/graphics/unaccelerated_static_bitmap_image.h"
 #include "third_party/blink/renderer/platform/heap/garbage_collected.h"
@@ -262,6 +263,10 @@ CanvasAsyncBlobCreator::CanvasAsyncBlobCreator(
                          std::min(info.height(), max_dimension));
       src_data_.reset(info, src_data_.addr(), src_data_.rowBytes());
     }
+    if (RuntimeEnabledFeatures::FingerprintingCanvasImageDataNoiseEnabled()) {
+      StaticBitmapImage::ShuffleSubchannelColorData(src_data_.writable_addr(),
+        src_data_.info(), 0, 0);
+    }
   }
 
   idle_task_status_ = kIdleTaskNotSupported;
--
2.25.1
```

- Hardening against incognito mode detection (fixup) 
#1615
```
From: Your Name <you@example.com>
Date: Wed, 31 Aug 2022 15:38:17 +0000
Subject: Hardening against incognito mode detection (fixup)

---
 storage/browser/quota/quota_settings.cc | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/storage/browser/quota/quota_settings.cc b/storage/browser/quota/quota_settings.cc
--- a/storage/browser/quota/quota_settings.cc
+++ b/storage/browser/quota/quota_settings.cc
@@ -28,7 +28,7 @@ const int64_t kMBytes = 1024 * 1024;
 const int kRandomizedPercentage = 10;
 const double kDefaultPerHostRatio = 0.75;
 const double kIncognitoQuotaRatioLowerBound = 0.15;
-const double kIncognitoQuotaRatioUpperBound = 0.3;
+const double kIncognitoQuotaRatioUpperBound = 0.2;
 
 // Skews |value| by +/- |percent|.
 int64_t RandomizeByPercent(int64_t value, int percent) {
@@ -60,7 +60,7 @@ absl::optional<QuotaSettings> CalculateNominalDynamicSettings(
   base::ScopedBlockingCall scoped_blocking_call(FROM_HERE,
                                                 base::BlockingType::MAY_BLOCK);
 
-  if (is_incognito) {
+  if ((false) && is_incognito) {
     return CalculateIncognitoDynamicSettings(
         device_info_helper->AmountOfPhysicalMemory());
   }
--
2.25.1
```

- Move navigation bar to bottom (fixup) 
#2253
```
From: Your Name <you@example.com>
Date: Wed, 31 Aug 2022 15:22:59 +0000
Subject: Move navigation bar to bottom (fixup)

---
 .../layouts/LayoutManagerChrome.java           | 18 +++++++++++++++---
 1 file changed, 15 insertions(+), 3 deletions(-)

diff --git a/chrome/android/java/src/org/chromium/chrome/browser/compositor/layouts/LayoutManagerChrome.java b/chrome/android/java/src/org/chromium/chrome/browser/compositor/layouts/LayoutManagerChrome.java
--- a/chrome/android/java/src/org/chromium/chrome/browser/compositor/layouts/LayoutManagerChrome.java
+++ b/chrome/android/java/src/org/chromium/chrome/browser/compositor/layouts/LayoutManagerChrome.java
@@ -45,6 +45,8 @@ import org.chromium.components.browser_ui.widget.gesture.SwipeGestureListener.Sc
 import org.chromium.components.browser_ui.widget.gesture.SwipeGestureListener.SwipeHandler;
 import org.chromium.components.browser_ui.widget.scrim.ScrimCoordinator;
 import org.chromium.ui.resources.dynamics.DynamicResourceLoader;
+import org.chromium.chrome.browser.flags.ChromeFeatureList;
+import org.chromium.chrome.browser.flags.CachedFeatureFlags;
 
 import java.util.List;
 
@@ -210,7 +212,10 @@ public class LayoutManagerChrome
 
     @Override
     public SwipeHandler createToolbarSwipeHandler(boolean supportSwipeDown) {
-        return new ToolbarSwipeHandler(supportSwipeDown);
+        boolean move_top_toolbar =
+            CachedFeatureFlags.isEnabled(ChromeFeatureList.MOVE_TOP_TOOLBAR_TO_BOTTOM);
+        return new ToolbarSwipeHandler(supportSwipeDown && !move_top_toolbar,
+                                       supportSwipeDown && move_top_toolbar);
     }
 
     @Override
@@ -424,9 +429,11 @@ public class LayoutManagerChrome
         private static final float SWIPE_RANGE_DEG = 25;
 
         private final boolean mSupportSwipeDown;
+        private final boolean mSupportSwipeUp;
 
-        public ToolbarSwipeHandler(boolean supportSwipeDown) {
+        public ToolbarSwipeHandler(boolean supportSwipeDown, boolean supportSwipeUp) {
             mSupportSwipeDown = supportSwipeDown;
+            mSupportSwipeUp = supportSwipeUp;
         }
 
         @Override
@@ -458,6 +465,9 @@ public class LayoutManagerChrome
                     && mScrollDirection == ScrollDirection.DOWN) {
                 RecordUserAction.record("MobileToolbarSwipeOpenStackView");
                 showLayout(LayoutType.TAB_SWITCHER, true);
+            } else if (mSupportSwipeUp && mOverviewLayout != null
+                    && mScrollDirection == ScrollDirection.UP) {
+                showLayout(LayoutType.TAB_SWITCHER, true);
             } else if (mScrollDirection == ScrollDirection.LEFT
                     || mScrollDirection == ScrollDirection.RIGHT) {
                 startShowing(mToolbarSwipeLayout, true);
@@ -504,6 +514,8 @@ public class LayoutManagerChrome
                 direction = ScrollDirection.RIGHT;
             } else if (swipeAngle < 270 + SWIPE_RANGE_DEG && swipeAngle > 270 - SWIPE_RANGE_DEG) {
                 direction = ScrollDirection.DOWN;
+            } else if (swipeAngle < 90 + SWIPE_RANGE_DEG && swipeAngle > 90 - SWIPE_RANGE_DEG) {
+                direction = ScrollDirection.UP;
             }
 
             return direction;
@@ -518,7 +530,7 @@ public class LayoutManagerChrome
                 return false;
             }
 
-            if (direction == ScrollDirection.DOWN) {
+            if (direction == ScrollDirection.DOWN || direction == ScrollDirection.UP) {
                 boolean isAccessibility = ChromeAccessibilityUtil.get().isAccessibilityEnabled();
                 return mOverviewLayout != null && !isAccessibility;
             }
--
2.25.1
```

**Comment by Pouriyarezaee** [comment]
😎

[Pouriya Rezaee - Chat @ Spike](https://spikenow.com/r/a/?ref=spike-organic-signature&_ts=1prrsn)	[1prrsn]

On September 5, 2022 at 11:42 GMT, uazo ***@***.***> wrote:

these are some fixups if you want to introduce them in the next build:

- Multiple fingerprinting mitigations (fixup)
fix the lack of protection in canvas.convertToBlob()
From: Your Name ***@***.***> Date: Fri, 2 Sep 2022 13:29:33 +0000 Subject: Multiple fingerprinting mitigations (fixup) --- .../renderer/core/html/canvas/canvas_async_blob_creator.cc | 5 +++++ 1 file changed, 5 insertions(+) diff --git a/third_party/blink/renderer/core/html/canvas/canvas_async_blob_creator.cc b/third_party/blink/renderer/core/html/canvas/canvas_async_blob_creator.cc --- a/third_party/blink/renderer/core/html/canvas/canvas_async_blob_creator.cc +++ b/third_party/blink/renderer/core/html/canvas/canvas_async_blob_creator.cc @@ -20,6 +20,7 @@ #include "third_party/blink/renderer/core/fileapi/blob.h" #include "third_party/blink/renderer/core/html/canvas/canvas_rendering_context.h" #include "third_party/blink/renderer/platform/graphics/image_data_buffer.h" +#include "third_party/blink/renderer/platform/graphics/static_bitmap_image.h" #include "third_party/blink/renderer/platform/graphics/skia/skia_utils.h" #include "third_party/blink/renderer/platform/graphics/unaccelerated_static_bitmap_image.h" #include "third_party/blink/renderer/platform/heap/garbage_collected.h" @@ -262,6 +263,10 @@ CanvasAsyncBlobCreator::CanvasAsyncBlobCreator( std::min(info.height(), max_dimension)); src_data_.reset(info, src_data_.addr(), src_data_.rowBytes()); } + if (RuntimeEnabledFeatures::FingerprintingCanvasImageDataNoiseEnabled()) { + StaticBitmapImage::ShuffleSubchannelColorData(src_data_.writable_addr(), + src_data_.info(), 0, 0); + } } idle_task_status_ = kIdleTaskNotSupported; -- 2.25.1
- Hardening against incognito mode detection (fixup)
[#1615](https://github.com/bromite/bromite/issues/1615)
From: Your Name ***@***.***> Date: Wed, 31 Aug 2022 15:38:17 +0000 Subject: Hardening against incognito mode detection (fixup) --- storage/browser/quota/quota_settings.cc | 4 ++-- 1 file changed, 2 insertions(+), 2 deletions(-) diff --git a/storage/browser/quota/quota_settings.cc b/storage/browser/quota/quota_settings.cc --- a/storage/browser/quota/quota_settings.cc +++ b/storage/browser/quota/quota_settings.cc @@ -28,7 +28,7 @@ const int64_t kMBytes = 1024 * 1024; const int kRandomizedPercentage = 10; const double kDefaultPerHostRatio = 0.75; const double kIncognitoQuotaRatioLowerBound = 0.15; -const double kIncognitoQuotaRatioUpperBound = 0.3; +const double kIncognitoQuotaRatioUpperBound = 0.2; // Skews |value| by +/- |percent|. int64_t RandomizeByPercent(int64_t value, int percent) { @@ -60,7 +60,7 @@ absl::optional<QuotaSettings> CalculateNominalDynamicSettings( base::ScopedBlockingCall scoped_blocking_call(FROM_HERE, base::BlockingType::MAY_BLOCK); - if (is_incognito) { + if ((false) && is_incognito) { return CalculateIncognitoDynamicSettings( device_info_helper->AmountOfPhysicalMemory()); } -- 2.25.1
- Move navigation bar to bottom (fixup)
[#2253](https://github.com/bromite/bromite/issues/2253)
From: Your Name ***@***.***> Date: Wed, 31 Aug 2022 15:22:59 +0000 Subject: Move navigation bar to bottom (fixup) --- .../layouts/LayoutManagerChrome.java | 18 +++++++++++++++--- 1 file changed, 15 insertions(+), 3 deletions(-) diff --git a/chrome/android/java/src/org/chromium/chrome/browser/compositor/layouts/LayoutManagerChrome.java b/chrome/android/java/src/org/chromium/chrome/browser/compositor/layouts/LayoutManagerChrome.java --- a/chrome/android/java/src/org/chromium/chrome/browser/compositor/layouts/LayoutManagerChrome.java +++ b/chrome/android/java/src/org/chromium/chrome/browser/compositor/layouts/LayoutManagerChrome.java @@ -45,6 +45,8 @@ import org.chromium.components.browser_ui.widget.gesture.SwipeGestureListener.Sc import org.chromium.components.browser_ui.widget.gesture.SwipeGestureListener.SwipeHandler; import org.chromium.components.browser_ui.widget.scrim.ScrimCoordinator; import org.chromium.ui.resources.dynamics.DynamicResourceLoader; +import org.chromium.chrome.browser.flags.ChromeFeatureList; +import org.chromium.chrome.browser.flags.CachedFeatureFlags; import java.util.List; @@ -210,7 +212,10 @@ public class LayoutManagerChrome @Override public SwipeHandler createToolbarSwipeHandler(boolean supportSwipeDown) { - return new ToolbarSwipeHandler(supportSwipeDown); + boolean move_top_toolbar = + CachedFeatureFlags.isEnabled(ChromeFeatureList.MOVE_TOP_TOOLBAR_TO_BOTTOM); + return new ToolbarSwipeHandler(supportSwipeDown && !move_top_toolbar, + supportSwipeDown && move_top_toolbar); } @Override @@ -424,9 +429,11 @@ public class LayoutManagerChrome private static final float SWIPE_RANGE_DEG = 25; private final boolean mSupportSwipeDown; + private final boolean mSupportSwipeUp; - public ToolbarSwipeHandler(boolean supportSwipeDown) { + public ToolbarSwipeHandler(boolean supportSwipeDown, boolean supportSwipeUp) { mSupportSwipeDown = supportSwipeDown; + mSupportSwipeUp = supportSwipeUp; } @Override @@ -458,6 +465,9 @@ public class LayoutManagerChrome && mScrollDirection == ScrollDirection.DOWN) { RecordUserAction.record("MobileToolbarSwipeOpenStackView"); showLayout(LayoutType.TAB_SWITCHER, true); + } else if (mSupportSwipeUp && mOverviewLayout != null + && mScrollDirection == ScrollDirection.UP) { + showLayout(LayoutType.TAB_SWITCHER, true); } else if (mScrollDirection == ScrollDirection.LEFT || mScrollDirection == ScrollDirection.RIGHT) { startShowing(mToolbarSwipeLayout, true); @@ -504,6 +514,8 @@ public class LayoutManagerChrome direction = ScrollDirection.RIGHT; } else if (swipeAngle < 270 + SWIPE_RANGE_DEG && swipeAngle > 270 - SWIPE_RANGE_DEG) { direction = ScrollDirection.DOWN; + } else if (swipeAngle < 90 + SWIPE_RANGE_DEG && swipeAngle > 90 - SWIPE_RANGE_DEG) { + direction = ScrollDirection.UP; } return direction; @@ -518,7 +530,7 @@ public class LayoutManagerChrome return false; } - if (direction == ScrollDirection.DOWN) { + if (direction == ScrollDirection.DOWN || direction == ScrollDirection.UP) { boolean isAccessibility = ChromeAccessibilityUtil.get().isAccessibilityEnabled(); return mOverviewLayout != null && !isAccessibility; } -- 2.25.1

**Comment by csagan5** [comment]
These seem bugfixes that require an entry in the CHANGELOG.

File: build/patches/Disable-third-party-origin-trials.patch [hunk 1]
```diff
@@ -159,25 +159,26 @@ diff --git a/third_party/blink/common/origin_trials/trial_token_validator.cc b/t
 diff --git a/third_party/blink/renderer/core/loader/document_loader.cc b/third_party/blink/renderer/core/loader/document_loader.cc
 --- a/third_party/blink/renderer/core/loader/document_loader.cc
 +++ b/third_party/blink/renderer/core/loader/document_loader.cc
-@@ -167,31 +167,18 @@ Vector<OriginTrialFeature> CopyInitiatorOriginTrials(
- WebVector<int> CopyInitiatorOriginTrials(
-     const Vector<OriginTrialFeature>& initiator_origin_trial_features) {
-   WebVector<int> result;
+@@ -159,13 +159,6 @@ constexpr base::FeatureParam<bool> kCacheInlineScriptCodeFixConfiguring{
+ Vector<OriginTrialFeature> CopyInitiatorOriginTrials(
```
Comment (uazo) [inline build/patches/Disable-third-party-origin-trials.patch:163]: there is another method (overload) to remove:

```
 WebVector<int> CopyInitiatorOriginTrials(
     const Vector<OriginTrialFeature>& initiator_origin_trial_features) {
   WebVector<int> result;
-  for (auto feature : initiator_origin_trial_features) {
-    // Convert from OriginTrialFeature to int. These values are passed between
-    // blink navigations. OriginTrialFeature isn't visible outside of blink (and
-    // doesn't need to be) so the values are transferred outside of blink as
-    // ints and casted to OriginTrialFeature once being processed in blink.
-    result.emplace_back(static_cast<int>(feature));
-  }
   return result;
 }
 
```

**Comment by chirayudesai** [comment]
It fails to build for me:

I applied these patches to 105.0.5195.41 and tried to build `chrome_public_apk` with the args.gn here. arm64, android

```
../../chrome/browser/commerce/price_tracking/android/price_tracking_notification_bridge.cc:9:10: fatal error: 'chrome/browser/commerce/price_tracking/android/jni_headers/PriceTrackingNotificationBridge_jni.h' file not found
```

**Comment by csagan5** [comment]
> It fails to build for me:
> 
> I applied these patches to 105.0.5195.41 and tried to build `chrome_public_apk` with the args.gn here. arm64, android
> 
> ```
> ../../chrome/browser/commerce/price_tracking/android/price_tracking_notification_bridge.cc:9:10: fatal error: 'chrome/browser/commerce/price_tracking/android/jni_headers/PriceTrackingNotificationBridge_jni.h' file not found
> ```

I suspect it does not fail for me because I first build Chromium and then Bromite, so it uses the Chromium headers.

**Pull request merged**